### PR TITLE
[Ong Xinyi] iP

### PR DESCRIPTION
**Pull request by ongxinyii**
### **Pookie**

Pookie 🎀 is your personal assistant who is _cute_, _loving_ and _helpful_, fit for a princess like yourself. 
> If I am a princess in rags and tatters, I can be a princess inside. Frances Hodgson Burnett, [A Little Princess](https://www.goodreads.com/work/quotes/1313599)

They are:
- text-based
- simple and functional

You can download Pookie from [here](https://github.com/ongxinyii/ip/releases).

**What's New**:
1. Enhanced Task Management features
2. Searching Tasks by Keyword
3. Enhanced UI (see Pookie's unique emotions!)

With `enum` Included:
```java
package pookie.task;

public enum TaskType {
    TODO,
    DEADLINE,
    EVENT
}
``` 

Checklist (future updates):
- [x] Added find feature
- [x] Wrote unit tests for edge cases
- [ ] Update documentation to reflect new feature
- [ ] Add GUI

### CODE REVIEW

File: src/main/java/pookie/command/Parser.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
+package pookie.command;
+
+import pookie.exception.PookieException;
+import pookie.list.TaskList;
+import pookie.storage.Storage;
+import pookie.task.*;
```
Comment (lilyium) [inline src/main/java/pookie/command/Parser.java]: It might be better to list imported classes explicitly.

File: src/main/java/pookie/command/Parser.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                    handleTodo(tasks, parts.length > 1 ? parts[1].trim() : "", ui, storage);
+                    break;
+                case DEADLINE:
+                    handleDeadline(tasks, parts.length > 1 ? parts[1].trim() : "", ui, storage);
+                    break;
+                case EVENT:
+                    handleEvent(tasks, parts.length > 1 ? parts[1].trim() : "", ui, storage);
+                    break;
+                default:
+                    throw new PookieException("Sowwieeee (╥﹏╥) Pookie doesn't understand...");
```
Comment (lilyium) [inline src/main/java/pookie/command/Parser.java]: Case clauses should not be indented.

File: src/main/java/pookie/command/Parser.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                case "E": // Event task
+                    if (parts.length < 5) throw new PookieException("Error: Missing event start and end time.");
+                    LocalDateTime startDate = LocalDateTime.parse(parts[3], DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+                    LocalDateTime endDate = LocalDateTime.parse(parts[4], DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+                    Event event = new Event(description, startDate, endDate);
+                    if (isDone) event.markDone();
+                    return event;
+
+                default:
+                    throw new PookieException("Error: Unknown task type.");
```
Comment (lilyium) [inline src/main/java/pookie/command/Parser.java]: Case clauses should not be indented.

File: src/main/java/pookie/list/TaskList.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
+package pookie.list;
+
+import pookie.task.*;
```
Comment (lilyium) [inline src/main/java/pookie/list/TaskList.java]: It might be better to list imported classes explicitly.

File: src/main/java/pookie/storage/Storage.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package pookie.storage;
+
+import pookie.task.*;
+import pookie.exception.PookieException;
+import pookie.command.Parser;
+
+import java.io.*;
```
Comment (lilyium) [inline src/main/java/pookie/storage/Storage.java]: It might be better to list imported classes explicitly.

File: src/main/java/pookie/task/Event.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+     * @param end         The end date/time of the event in "yyyy-MM-dd HHmm" format.
+     * @throws PookieException If the date/time format is incorrect.
+     */
+    public Event(String description, String start, String end) throws PookieException {
+        super(description);
+        try {
+            this.start = LocalDateTime.parse(start, INPUT_FORMAT);
+            this.end = LocalDateTime.parse(end, INPUT_FORMAT);
+        } catch (DateTimeParseException e) {
+            throw new PookieException("Princess, please enter the event times in the correct format: yyyy-MM-dd HHmm (e.g., 2019-12-02 1800).");
```
Comment (lilyium) [inline src/main/java/pookie/task/Event.java]: You might want to consider wrapping this line.

**Review by lilyium — COMMENTED** [review]
LGTM :+1:! You might want to try to have the import statements in a consistent order throughout the files.

File: src/main/java/pookie/command/Parser.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            }
+        } catch (Exception e) {
+            throw new PookieException("Error parsing task: " + e.getMessage());
+        }
+    }
+
+    /**
+     * Handles creation of a ToDo task.
+     */
+    private static void handleTodo(TaskList tasks, String description, Ui ui, Storage storage) throws PookieException {
```
Comment (boinkkitty) [inline src/main/java/pookie/command/Parser.java:370]: Could consider renaming to handleToDo to maintain consistency with ToDo class name

File: src/main/java/pookie/command/Parser.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+    /**
+     * Handles creation of a Deadline task.
+     */
+    private static void handleDeadline(TaskList tasks, String details, Ui ui, Storage storage) throws PookieException {
+        String[] parts = details.split(" /by ");
+        if (parts[0].trim().isEmpty()) {
+            throw new PookieException.EmptyDescriptionException("Princess, the description of a deadline cannot be empty.");
+        }
+        if (parts.length < 2 || parts[1].trim().isEmpty()) {
```
Comment (boinkkitty) [inline src/main/java/pookie/command/Parser.java:388]: I think coding convention requires it to be in an else if and else block

File: src/main/java/pookie/command/Parser.java [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            ui.showError("OOPS!!! Princess, please enter the deadline in the correct format: yyyy-MM-dd HHmm (e.g., 2025-01-31 1500).");
+        }
+    }
+
+    /**
+     * Handles creation of an Event task.
+     */
+    private static void handleEvent(TaskList tasks, String details, Ui ui, Storage storage) throws PookieException {
+        String[] parts = details.split(" /from | /to ");
+        if (parts[0].trim().isEmpty()) {
```
Comment (boinkkitty) [inline src/main/java/pookie/command/Parser.java]: I think coding convention requires it to be in an else if and else block

File: src/main/java/pookie/command/Parser.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+            tasks.addTask(eventTask, ui, storage);
+        } catch (DateTimeParseException e) {
+            ui.showError("⚠️ OOPS!!! Princess, please enter event times in the correct format: yyyy-MM-dd HHmm (e.g., 2025-02-01 1800).");
+        }
+    }
+
+    /**
+     * Lists tasks occurring on a specific date.
+     */
+    private static void handleListByDate(TaskList tasks, String dateStr, Ui ui) {
```
Comment (boinkkitty) [inline src/main/java/pookie/command/Parser.java:441]: Cool method to search by date

File: src/main/java/pookie/exception/PookieException.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
+package pookie.exception;
+
+public class PookieException extends Exception {
+    public PookieException(String message) {
+        super(message);
+    }
+
```
Comment (boinkkitty) [inline src/main/java/pookie/exception/PookieException.java:20]: I think Javadoc comments should be in /** format

File: src/main/java/pookie/task/Task.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        this.isDone = false;
+    }
+
+    /**
+     * Converts the task into a format suitable for saving to a file.
+     * This method must be implemented by subclasses.
+     *
+     * @return A formatted string representation of the task.
+     */
+    public abstract String toFileFormat();
```
Comment (boinkkitty) [inline src/main/java/pookie/task/Task.java:55]: Nice use of abstract class and method

File: src/main/java/pookie/task/TaskType.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package pookie.task;
+
+public enum TaskType {
```
Comment (boinkkitty) [inline src/main/java/pookie/task/TaskType.java:13]: Nice use of enum

File: src/main/java/pookie/ui/Ui.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    public void close() {
+        sc.close();
+    }
+
+    /**
+     * Prints a horizontal border line for message separation.
+     */
+    private void printBorder() {
+        System.out.println("____________________________________________________________");
```
Comment (boinkkitty) [inline src/main/java/pookie/ui/Ui.java:83]: Could consider initializing as a constant like 'DIVIDER' if you prefer

**Review by boinkkitty — COMMENTED** [review]
LGTM. Just minor things like packaging and if else blocks.

**Comment by meglwhy** [comment]
LGTM! just minor nits but coding style is good
